### DISCUSSION TITLE
Blazor Component Libraries

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#2029] [General]
The default Oqtane administration experience relies on native Bootstrap 5, however the philosophy in Oqtane is to provide the flexibility for developers to use their preferred Blazor component library for custom application development. That being said, one of the most common questions asked in this forum is how to integrate various Blazor component libraries with Oqtane. The challenge is that each Blazor component library has their own unique requirements for setup in Blazor... including DLLs, static resources such as stylesheets and JavaScript, service registrations during startup, and components which need to be included in every page.

In order to simplify this effort, I am considering creating a number of Oqtane "modules" which would simplify the installation process for Blazor component libraries. For example, if someone wanted to use MudBlazor they could simply install the MudBlazor module into their Oqtane installation and it would include all of the necessary configuration. Once installed, developers could then use MudBlazor components in their custom modules.

Does this sound like something which developers would find useful?

**Comment by curlyfro** [comment]
i'm not too sure this is a good idea.  sounds like a maintenance nightmare.  so if I wanted a syncfusion module, matblazor, or fluxor, you'd create these modules for me?  isn't this just a matter of developers knowing how to theme better?  or are you saying that you're going to abstract out the default styling dependencies (Bootstrap 5) and make them optional modules?

**Reply by sbwalker** [reply]
I guess I did not describe this very well...

The only purpose of the component library module would be to install all of the resources for the component library and register them so that a module developer could use the components when developing their own custom modules. They would also be open source so would be a useful reference for developers. 

These modules would be completely optional. If a developer wants to figure out the steps to manually integrate a component library they absolutely could.

**Reply by curlyfro** [reply]
i'm sorry Shaun i still don't understand.  the idea of themes/modules/plugins exist is all CMS systems -- I get that.  you're saying you're introducing another layer for 3rd party components?  what can't theme/modules do that this component library can do? or vice versa?  if there's a disconnect there then maybe the module registration should get changed.

i just looked at your code and you have IClientStartup and I think netcore has IStartupFilter that users can use for service registration.  what's the problem?

**Reply by sbwalker** [reply]
Oqtane is not a CMS - it is a modular application framework. Extensions in Oqtane are simply Nuget packages. A Nuget package can contain any combination of themes, modules, assemblies, startup classes, and static resources. So what I am talking about here is a standard Oqtane Nuget package. 

The problem I am trying to solve with a component library package is that it is complicated to integrate many of the popular component libraries into Oqtane because they have many different resources which need to be deployed in specific locations and services which need to be registered. I would encourage you to try to configure a library yourself in Oqtane to understand the challenges. As a result there are many threads in this forum by developers asking for help in configuring component libraries. So I am suggesting that an installable package could potentially make this much easier. 

Hopefully this provides a clearer explanation.

**Reply by DrtyByt** [reply]
Honestly after struggling with this exact thing for the last week I think this would be amazing! If I understand right this is basically a Module that we add to the Oqtane modules and then to ingest/use that component library all we need to do in our Modules that we want to use it is add the assembly reference and then it essence it just works instead of us trying to configure all the dependencies needed for the component library to work.

**Reply by sbwalker** [reply]
@DrtyByt exactly

**Reply by ADefWebserver** [reply]
@sbwalker - My primary concern is that someone installs my module and it's working fine, then they upgrade the "common 3rd part component package" and my module stops working because I did not support the breaking changes in that package.

**Reply by sbwalker** [reply]
That problem could happen regardless of whether a "component module" package existed or not. If 2 module developers use the same component library and that component library has incompatibilities then it could cause issues when the modules are installed into an Oqtane installation. The existence of a "component module" package neither helps nor hinders this problem.

**Reply by ADefWebserver** [reply]
@sbwalker - Yeah the .dll would be overwritten. The .js/.css is currently "per module" when using code like this:

```
            new Resource {
                ResourceType = ResourceType.Script,
                Url = "Modules/" + GetType().Namespace 
                + "/fonts/MaterialIcons-Regular.woff" },
            new Resource {
                ResourceType = ResourceType.Stylesheet,
                Url = "Modules/" + GetType().Namespace 
                + "/css/default.css" },
            new Resource {
                ResourceType = ResourceType.Script,
                Url = "Modules/" + GetType().Namespace 
                + "/js/Radzen.Blazor.js" },
```
However, a module developer could still keep using this method I guess.

**Comment by Rodien** [comment]
I think that this would definitely improve the developer experience🚀
